File: leetcode/92-reverse-linked-list-ii/main.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode

        1st approach
        - similar to reverse linked list and reverse halh of a linked list
        - we need to find the parent of node[m] first, reverse the later list until the count < n(right)
        20ms beats 100%
        29jan2019
        """
        # find the parent of node[m]
        dumphead = ListNode(0)
        dumphead.next = head

        parent = dumphead
        cur = head
        count = 1
        while count < m and cur != None:
            parent = cur
            cur = cur.next
            count += 1
        # reverse the linked list until n
        newHead = cur
        while cur != None and cur.next != None and count < n:
            temp = cur.next
            cur.next = cur.next.next
            temp.next = newHead
            newHead = temp
            count += 1
        parent.next = newHead
        return dumphead.next


# helpers for test
def arr2list(arr):
    dump = ListNode(0)
    dumpCur = dump
    for i in range(len(arr)):
        dumpCur.next = ListNode(arr[i])
        dumpCur = dumpCur.next
    return dump.next


def printList(l):
    c = l
    acc = ""
    while c != None:
        acc += str(c.val)+','
        c = c.next
    print(acc)


a = arr2list([])
s = Solution().reverseBetween(a, 0, 0)
printList(s)

a = arr2list([1])
s = Solution().reverseBetween(a, 1, 1)
printList(s)

a = arr2list([1, 2, 3, 4, 5])
s = Solution().reverseBetween(a, 2, 4)
printList(s)

a = arr2list([1, 2, 3, 4, 5])
s = Solution().reverseBetween(a, 2, 5)
printList(s)

a = arr2list([1, 2, 3, 4, 5])
s = Solution().reverseBetween(a, 1, 4)
printList(s)

a = arr2list([1, 2, 3, 4, 5])
s = Solution().reverseBetween(a, 1, 5)
printList(s)

File: leetcode/92-reverse-linked-list-ii/main.go
package main

import (
	"fmt"
	"strconv"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseBetween(head *ListNode, m int, n int) *ListNode {
	if head == nil {
		return nil
	}
	// find the parent of node[m]
	dump := &ListNode{0, head}
	i := 1
	parent := dump
	for i < m && parent != nil {
		i++
		parent = parent.Next
	}
	// reverse the linked list until n
	targetHead := parent.Next
	cur := targetHead
	for i < n && targetHead.Next != nil {
		temp := targetHead.Next
		targetHead.Next = targetHead.Next.Next
		temp.Next = cur
		cur = temp
		i++
	}
	parent.Next = cur
	return dump.Next
}

// helpers
func arr2list(arr []int) *ListNode {
	dump := &ListNode{0, nil}
	dumpCur := dump
	for i := 0; i < len(arr); i++ {
		dumpCur.Next = &ListNode{arr[i], nil}
		dumpCur = dumpCur.Next
	}
	return dump.Next
}

func printList(l *ListNode) {
	c := l
	s := ""
	for c != nil {
		s += strconv.Itoa(c.Val) + ","
		c = c.Next
	}
	fmt.Println(s)
}

func main() {

	a := arr2list([]int{})
	b := reverseBetween(a, 0, 0)
	printList(b)

	a = arr2list([]int{1})
	b = reverseBetween(a, 1, 1)
	printList(b)

	a = arr2list([]int{1, 2, 3, 4, 5})
	b = reverseBetween(a, 2, 4)
	printList(b)

	a = arr2list([]int{1, 2, 3, 4, 5})
	b = reverseBetween(a, 2, 5)
	printList(b)

	a = arr2list([]int{1, 2, 3, 4, 5})
	b = reverseBetween(a, 1, 4)
	printList(b)

	a = arr2list([]int{1, 2, 3, 4, 5})
	b = reverseBetween(a, 1, 5)
	printList(b)
}
